src/libmw/src/crypto/Schnorr.cpp

```diff
@@ -0,0 +1,138 @@
+#include <mw/crypto/Schnorr.h>
+#include "Context.h"
+#include "ConversionUtil.h"
+#include "PublicKeys.h"
+#include "SchnorrCache.h"
+
+#include <mw/common/Logger.h>
+#include <mw/exceptions/CryptoException.h>
+#include <mw/util/VectorUtil.h>
+
+static SchnorrCache CACHE;
+static Locked<Context> SCHNORR_CONTEXT(std::make_shared<Context>());
+
+static constexpr uint64_t MAX_WIDTH = 1 << 20;
+static constexpr size_t SCRATCH_SPACE_SIZE = 256 * MAX_WIDTH;
+
+Signature Schnorr::Sign(
+    const uint8_t* secretKey,
+    const mw::Hash& message)
+{
+    secp256k1_schnorrsig signature;
+    const int signedResult = secp256k1_schnorrsig_sign(
+        SCHNORR_CONTEXT.Write()->Randomized(),
+        &signature,
+        nullptr,
+        message.data(),
+        secretKey,
+        nullptr,
+        nullptr
+    );
+    if (signedResult != 1) {
+        ThrowCrypto("Failed to sign message.");
+    }
+
+    return ConversionUtil(SCHNORR_CONTEXT).ToSignature(signature);
+}
+
+SignedMessage Schnorr::SignMessage(
+    const BigInt<32>& secretKey,
+    const mw::Hash& message)
+{
+    PublicKey pubkey = PublicKeys(SCHNORR_CONTEXT).CalculatePublicKey(secretKey);
+    Signature sig = Sign(secretKey.data(), message);
+
+    return SignedMessage(message, pubkey, sig);
+}
+
+bool Schnorr::Verify(
+    const Signature& signature,
+    const PublicKey& sumPubKeys,
+    const mw::Hash& message)
+{
+    SignedMessage signed_message(message, sumPubKeys, signature);
+    if (CACHE.Contains(signed_message)) {
+        return true;
+    }
+
+    secp256k1_pubkey parsedPubKey = ConversionUtil(SCHNORR_CONTEXT).ToSecp256k1(sumPubKeys);
+
+    const int verifyResult = secp256k1_aggsig_verify_single(
+        SCHNORR_CONTEXT.Read()->Get(),
+        signature.data(),
+        message.data(),
+        nullptr,
+        &parsedPubKey,
+        &parsedPubKey,
+        nullptr,
+        false
+    );
+
+    if (verifyResult == 1) {
+        CACHE.Add(signed_message);
+    }
+
+    return verifyResult == 1;
+}
+
+bool Schnorr::BatchVerify(const std::vector<SignedMessage>& signatures)
+{
+    std::vector<SignedMessage> messages;
+    for (const SignedMessage& signed_message : signatures) {
+        if (!CACHE.Contains(signed_message)) {
+            messages.push_back(signed_message);
+        }
+    }
+
+    // Parse pubkeys
+    std::vector<secp256k1_pubkey> parsedPubKeys;
+    std::transform(
```

Comment:
At the moment there are 4 loops over signatures. Can we just have a single loop?